### PR TITLE
Make Featured Product Review Data More Accurate

### PR DESCRIPTION
**Pull request by danielcliu-ifixit**
## Overview
For the featured product sections at the bottom of a product page, we weren't always getting the most up to date data in terms of reviews. This change fixes that by having the featured products make calls to the ifixit api for review data instead of using the review data from shopify (where possible).

## CR
I'm not an expert in async, so I'm unsure if this is the best solution. I originally tried to use `useProductReviews()` but the new data never seemed to get fetched/never updated the existing data anyways, so I just use the api fetch.

## QA

Make sure that the issues on `/products/2-tb-ssd-hybrid-3-5-hard-drive` mentioned in the issue are no longer there.

Closes: #859

### CODE REVIEW

**Comment by jordycosta** [comment]
QA 🟢 

Products in the featured products section now have accurate review counts 👍 
[Product page tested](https://react-commerce-git-sync-review-count-on-feature-products-ifixit.vercel.app/products/2-tb-ssd-hybrid-3-5-hard-drive)

ratings

---

pro-tech

mako

File: frontend/templates/product/sections/FeaturedProductsSection/index.tsx [hunk 1]
```diff
@@ -108,13 +116,43 @@ function ProductGridItem({
...
+         sku
+      );
+      return reviewsData;
+   };
+
+   const [reviewsData, initReviews] = useState<
+      ProductReviewData | null | undefined
+   >();
+   useEffect(() => {
+      fetchData()
```
Comment (masonmcelvain) [inline frontend/templates/product/sections/FeaturedProductsSection/index.tsx:143]: This is a little briefer and obviates the need for the `fetchData` callback

```suggestion
      if (!sku) {
         return;
      }
      fetchProductReviews(appContext.ifixitOrigin, sku)
```

File: frontend/templates/product/sections/FeaturedProductsSection/index.tsx [hunk 2]
```diff
@@ -108,13 +116,43 @@ function ProductGridItem({
...
+         return null;
+      }
+      const reviewsData = await fetchProductReviews(
+         appContext.ifixitOrigin,
+         sku
+      );
+      return reviewsData;
+   };
+
+   const [reviewsData, initReviews] = useState<
```
Comment (masonmcelvain) [inline frontend/templates/product/sections/FeaturedProductsSection/index.tsx:139]: Generally speaking, I think it's [convention](https://reactjs.org/docs/hooks-state.html#declaring-a-state-variable) to name the state update function like `setNameOfStateVar`. I think it makes it a little clearer later on that you're updating React state, as opposed to invoking some other kind of callback.

So here, I would personally name this
```suggestion
   const [reviewsData, setReviewsData] = useState<
```

File: frontend/templates/product/sections/FeaturedProductsSection/index.tsx [hunk 3]
```diff
@@ -23,9 +23,15 @@ import {
...
-import { Product } from '@models/product';
+import {
+   fetchProductReviews,
+   Product,
+   ProductReviewData,
+} from '@models/product';
 import { ImagePlaceholder } from '@templates/product/components/ImagePlaceholder';
 import NextLink from 'next/link';
+import { useAppContext } from '@ifixit/app';
+import React, { useState, useEffect } from 'react';
```
Comment (masonmcelvain) [inline frontend/templates/product/sections/FeaturedProductsSection/index.tsx:34]: I think in this repo we've been following the pattern of
```suggestion
import * as React from 'react';
```

And then using the hooks like `React.useState` and `React.useEffect`. It probably doesn't matter, but just thought I'd share the pattern we've been using.

File: frontend/templates/product/sections/FeaturedProductsSection/index.tsx [hunk 4]
```diff
@@ -108,13 +116,43 @@ function ProductGridItem({
...
+   >();
+   useEffect(() => {
+      fetchData()
+         .then((res) => {
+            initReviews(res);
+         })
+         .catch(() => {
+            initReviews(null);
+         });
+   }, []);
```
Comment (masonmcelvain) [inline frontend/templates/product/sections/FeaturedProductsSection/index.tsx:150]: Generally, the dependency array should include any local vars/props that are used in `useEffect`, though if ESLint didn't complain maybe it's not a problem here.

**Review by masonmcelvain — CHANGES_REQUESTED** [review]
CR ✌🏻 Nice job glad it works!

I might have missed the conversation, but did we consider re-pushing product data to Shopify whenever a product review is made (in iFixit/ifixit) as opposed to making more api requests on every product page view?

dev_block ✌🏻 on that and a few other questions

**Comment by danielbeardsley** [comment]
> I might have missed the conversation, but did we consider re-pushing product data to Shopify whenever a product review is made (in iFixit/ifixit) as opposed to making more api requests on every product page view?

I really think this is preferable. It's like a 1 or 2-line change to our php app and would mean this whole pull isn't needed and we wouldn't need to make additional api requests.

**Comment by danielcliu-ifixit** [comment]
Closing for now in favor of https://github.com/iFixit/ifixit/pull/45303 which should fix the issue by updating shopify whenever a review is made.

**Pull request closed without merging**